wandb/sdk/internal/internal_api.py

```diff
@@ -3539,47 +3539,97 @@ def _resolve_org_entity_name(self, entity: str, organization: str = "") -> str:
         # the organization parameter, or an error if it is empty. Otherwise, this returns the
         # fetched value after validating that the given organization, if not empty, matches
         # either the org's display or entity name.
+
+        if not entity:
+            raise ValueError("Entity name is required to resolve org entity name.")
+
         org_fields = self.server_organization_type_introspection()
-        can_fetch_org_entity = "orgEntity" in org_fields
-        if not organization and not can_fetch_org_entity:
+        can_shorthand_org_entity = "orgEntity" in org_fields
+        if not organization and not can_shorthand_org_entity:
             raise ValueError(
                 "Fetching Registry artifacts without inputting an organization "
                 "is unavailable for your server version. "
                 "Please upgrade your server to 0.50.0 or later."
             )
-        if not can_fetch_org_entity:
+        if not can_shorthand_org_entity:
             # Server doesn't support fetching org entity to validate,
             # assume org entity is correctly inputted
             return organization
 
-        org_entity, org_name = self.fetch_org_entity_from_entity(entity)
+        orgs_from_entity = self._fetch_orgs_and_org_entities_from_entity(entity)
         if organization:
-            if organization != org_name and organization != org_entity:
-                raise ValueError(
-                    f"Artifact belongs to the organization {org_name!r} "
-                    f"and cannot be linked/fetched with {organization!r}. "
-                    "Please update the target path with the correct organization name."
+            return self._match_org_with_fetched_org_entities(
+                organization, orgs_from_entity
+            )
+
+        # If no input organization provided, error if entity belongs to multiple orgs because we
+        # cannot determine which one to use.
+        if len(orgs_from_entity) > 1:
+            raise ValueError(
+                f"Personal entity {entity!r} belongs to multiple organizations "
+                "and cannot be used without specifying the organization name. "
+                "Please specify the organization in the Registry path or use a team entity in the entity settings."
+            )
+        return orgs_from_entity[0][0]
+
+    def _match_org_with_fetched_org_entities(
+        self, organization: str, orgs: List[Tuple[str, str]]
+    ) -> str:
```

Comment:
suggestion: this can be pulled out as a module-level function, instead of being bound to `InternalApi`